packages/shared/lib/core/profile-manager/api/events/transaction-progress.event.ts

```diff
@@ -0,0 +1,83 @@
+import { selectedAccountId } from '@core/account'
+import { get } from 'svelte/store'
+import { isLedgerProfile } from '@core/profile'
+import { closePopup, openPopup } from '@lib/popup'
+import { ledgerDeviceStatus, ledgerMintNativeTokenConfirmationProps, ledgerSendConfirmationProps } from '@core/ledger'
+
+import { TransactionProgressPayload } from '../types'
+
+export function handleTransactionProgressEvent(accountId: string, payload: TransactionProgressPayload): void {
+    if (get(isLedgerProfile)) {
+        if (get(selectedAccountId) === accountId) {
+            handleTransactionProgressInternal(payload)
+        }
+    } else {
+        console.warn('Transaction progress handler unimplemented: ', payload)
+    }
+}
+
+function handleTransactionProgressInternal(
+    payload: TransactionProgressPayload,
+    isDuringOnboarding: boolean = false
+): void {
+    const sendConfirmationProps = get(ledgerSendConfirmationProps)
+    const mintNativeTokenConfirmationProps = get(ledgerMintNativeTokenConfirmationProps)
+
+    const _openPopup = (type = 'ledgerTransaction') => {
+        const shouldPreventClose = type === 'enableLedgerBlindSigning' || type === 'ledgerTransaction'
+
+        const isBlindSigningTransaction = typeof payload !== 'string' && 'PreparedTransactionEssenceHash' in payload
+        const toAddress = isBlindSigningTransaction
+            ? null
+            : sendConfirmationProps.recipient.type === 'address'
+            ? sendConfirmationProps.recipient.address
+            : sendConfirmationProps.recipient.account.depositAddress
+        const toAmount = isBlindSigningTransaction
+            ? null
+            : `${sendConfirmationProps.amount} ${sendConfirmationProps.unit}`
```

Comment:
This style of coding is very hard to follow/debug. The following style makes it more clear:
I would extract functions here, since this is only relevant information if you want to how the `toAddress` is derived, but not if you want to understand how `handleTransactionProgressInternal` is working.

```
function getToAddress(subject: Subject): string {
  if (isBlindSigningTransaction) {
    return null
  }
  if (...) {
    return subject.address
  }
  ...
}
```